WordPress/openverse
# Add ESLint plugin package

## Fixes



Fixes #2366 by @sarayourfriend

## Description


This PR adds a new `packages` directory with a single new package, `@openverse/eslint-plugin`. It configures various things to support the packages:

- `tsconfig.json` is configured to be a monolithic configuration. This is necessary because `vue-tsc` does not support composite projects and even if it did, it would require emitting definition files which we don't need. Using a monolithic configuration is faster in IDEs and in type checking. This required making changes to places that assumed the tsconfig lived in the frontend directory. I ran into difficulties here because I made erroneous changes to the tsconfig (because I originally _tried_ to use a composite configuration, which vue-tsc doesn't support). These led to a lot of debugging and back and forth in this PR that you'll see in the comments below. However! Do not let that make it seem like this change is more complex than it is. Both the frontend unit tests breaking and the bundle size increases were caused by me incorrectly (and unintentionally) changing tsconfig settings that led to changes in how babel and webpack were transpiling files. 
- Adds new CI/CD checks to run package build and unit tests
- Adds `eslint` as a script to the root `package.json` to facilitate running the script with and without `--fix`

The diff stats look huge, but `pnpm-lock.yaml` accounts for ~ 2,653 changes: 816 additions & 1,837 deletions (i.e., the _vast_ majority of them). In reality there is very little _changing_ here other than the TypeScript configuration changes (described above). Almost all of this PR is aimed at adding the packages directory, CI/CD checks for those packages, and the new eslint plugin package along with tests and documentation.

The new eslint plugin has a single rule to check the analytics configuration and encodes the following assumptions:

1. Payload values must be numbers, boolean values, or strings
2. Events with no payload must still be configured on the `Events` type with a `never` payload type
3. Events must be named in screaming snake case
4. Events must not include payload members with keys overlapping the default keys included for all events

> **Note**
>
> I tried to think of ways to automatically configure the default props but couldn't think of any that made much sense. If you have any ideas, let me know!

> **Note**
>
> There are limitations to what the rule can do because it does not have access to type information. Therefore, it allows any type alias/reference, even if it could potentially be to an invalid type. For example, if a type `type Whatever = { foo: 'bar' }` existed, this would not be a valid payload prop value in practice, but `EVENT: { whatever: Whatever }` would pass the linting check. The TypeScript ESLint plugin does have a feature that allows you to write rules that consume type information and it would be nice to enable it one day, but it adds significant time to linting (TS needs to build the project first) and our project has heaps of errors generated by the recommended configuration of it that would need to be fixed first. These limitations mean we still need to be thoughtful when reviewing PRs that add analytics events. The rule aims to catch the vast majority of accidental misconfiguration.

## Testing Instructions

Run `pnpm --filter eslint-plugin run test:unit` to confirm the unit tests pass (though they should also pass explicitly in the new CI/CD check).

Run `pnpm run eslint` to run ESLint with the new rule configured. This will automatically run the build for the eslint-plugin package (required because ESLint only supports rules in plain JS and the package is written in TypeScript).

Make changes to `types/analytics.ts`'s `Events` type that you believe should be valid and confirm that there are no false positive errors. Make changes to the type that should not be valid and confirm that those do indeed raise an error.

## Checklist


- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
